### DISCUSSION TITLE
Build bottle for versionned Formulae in own taps (bottle custom name ?)

### DISCUSSION THREAD

**Discussion by Laucans** [Homebrew/discussions#3899] [Writing Formulae/Casks]
Hello everyone, 

First message here

## Context 

I'm working for [LIGO](https://ligolang.org/ ) which is a smart contract language for [tezos](https://tezos.com/) and for mac users we distribute our language through Homebrew. 

For now we manage our tap (planned to migrate into homebrew-core when we will be more convenient with homebrew).

Currently if you want to install ligo compiler (after run `brew tap ligolang/ligo https://gitlab.com/ligolang/ligo.git`) you can run :
- `brew install ligolang/ligo/ligo` for the current version of ligo
- `brew install ligolang/ligo/ligo@X.Y.Z` for a specific version of ligo (we are managing 3 latest versions) 

Theses command build ligo on your machine following [install instructions described in formulae](https://gitlab.com/ligolang/ligo/-/blob/dev/HomebrewFormula/ligo.rb#L27)

## Bottle setup

But time to install is long, that's why we would like to implements bottle. (don't check current bottle setup, it's just false.) 
Working on it and I want to use [gitlab generic package registry](https://docs.gitlab.com/ee/user/packages/generic_packages/) to store our bottles. 
Everything works for the current version `brew install ligolang/ligo/ligo`
- A package named  `ligo-0.55.0.arm64_monterey.bottle.1.tar.gz` has been built and saved into [registry](https://gitlab.com/ligolang/ligo/-/packages/10531766)
- A root_url is configured to `https://gitlab.com/api/v4/projects/12294987/packages/generic/ligo_bottle/current/`

When I ran `brew install` the bottle is downloaded and installed.

(
Little parenthesis here, build the bottle with `brew bottle` generate a tar.gz name with 2 dash after the formulae name `ligo--0.55.0.arm64_monterey.bottle.1.tar.gz` but when I ran brew install the downloaded package got only one dash `ligo-0.55.0.arm64_monterey.bottle.1.tar.gz` Something wrong with my formulae ? There is a way to brew bottle with custom name ? to bypass this issue, I rename the file before upload it into the registry
)

## Issue

But if I want to do the same for a versioned formulae for example [ligo@0.55.0](https://gitlab.com/ligolang/ligo/-/blob/dev/HomebrewFormula/ligo%400.55.0.rb) the generated package is `ligo@0.55.0--0.55.0.arm64_monterey.bottle.1.tar.gz`

Here it contains `@` char which cannot be uploaded to [gitlab generic package repository](https://docs.gitlab.com/ee/user/packages/generic_packages/#publish-a-package-file)

The name follow a [format](https://docs.brew.sh/Bottles#format) which is not explicit in doc. 

> file_name	string	yes	The filename. It can contain only lowercase letters (a-z), uppercase letter (A-Z), numbers (0-9), dots (.), hyphens (-), or underscores (_).

Also when we want to install the version 0.55.0 artifact name is taken from the formula name (with @ converted into %40) instead of the ruby class the artifact downloaded is https://gitlab.com/api/v4/projects/12294987/packages/generic/ligo_bottle/current/ligo%400.55.0-0.55.0.arm64_monterey.bottle.1.tar.gz 

Which can't exist on gitlab du to the naming convention. 

Is there a way to manage bottle name in formulae through a custom name to avoid registry constraint ? 

Thanks

**Comment by ljames8** [comment]
Hi,
First, thanks for your post because it helped me setup my gitlab powered private tap, and successfully support bottles for the latest version.

What I ended up doing to support bottles for previous versions is working around by dropping the `<formula_name>@<some_previous_version>` convention and simply rename the older version formula with something like `ligo-0.55.0.rb`  in your case (with a hyphen instead of '@'). Accordingly the formula class name had to become `Ligo0550` rather than `LigoAT0550` . It seems to work.
